### PR TITLE
Fix `panic` message for `BTreeSet`'s `range` API and document `panic` cases

### PR DESCRIPTION
**Pull request by tnballo**
Currently, the `panic` cases for [`BTreeSet`'s `range` API](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#method.range) are undocumented and produce a slightly wrong `panic` message (says `BTreeMap` instead of `BTreeSet`).

Panic case 1 code:

```rust
use std::collections::BTreeSet;
use std::ops::Bound::Excluded;

fn main() {
    let mut set = BTreeSet::new();
    set.insert(3);
    set.insert(5);
    set.insert(8);

    for &elem in set.range((Excluded(&3), Excluded(&3))) {
        println!("{elem}");
    }
}
```

Panic case 1 message:

```
thread 'main' panicked at 'range start and end are equal and excluded in BTreeMap', /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/collections/btree/search.rs:105:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Panic case 2 code:

```rust
use std::collections::BTreeSet;
use std::ops::Bound::Included;

fn main() {
    let mut set = BTreeSet::new();
    set.insert(3);
    set.insert(5);
    set.insert(8);

    for &elem in set.range((Included(&8), Included(&3))) {
        println!("{elem}");
    }
}
```

Panic case 2:

```
thread 'main' panicked at 'range start is greater than range end in BTreeMap', /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/collections/btree/search.rs:110:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This PR fixes the output messages to say `BTreeSet`, adds the relevant unit tests, and updates the documentation for the API.

### CODE REVIEW

File: library/alloc/src/collections/btree/search.rs [hunk 1]
```diff
@@ -97,17 +97,30 @@ impl<BorrowType: marker::BorrowType, K, V> NodeRef<BorrowType, K, V, marker::Lea
         K: Borrow<Q>,
         R: RangeBounds<Q>,
     {
+        // A set's value is the unit type, ().
+        // `core::any::TypeId::of::<()>() == core::any::TypeId::of::<V>()` requires `V: 'static`.
+        // This workaround takes advantage of `()` being a zero-sized type.
+        let is_set = core::mem::size_of::<()>() == core::mem::size_of::<V>();
```
Comment (compiler-errors) [inline library/alloc/src/collections/btree/search.rs:103]: Won't this print `BTreeSet` for any BTreeMap with ZST value then?

Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: Unfortunately yes! Still believe it's an improvement over the current state of printing `BTreeMap` for `BTreeSet` all the time because ZSTs should be relatively rare.

Is there a way to get this 100% accurate? 

Wanted to avoid adding a private "tag" field to both collection structs, that was the only other option I could think of.

Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: Other idea: if it's appropriate to truncate the messages (e.g. remove `in BTreeMap` from both) then there's no need to determine which container it is, and the message is always right without additional logic.

Comment (compiler-errors) [reply library/alloc/src/collections/btree/search.rs:103]: Hm.. yeah, I guess you're limited by what is const, so even checking if the value type is `()` is difficult to do here. I wish there were a `type_id` intrinsic that ignored lifetimes...

I'm not on t-libs so this is just my personal opinion, but maybe the "in BTreeSet" is unnecessary, yeah.

Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: Agreed, a more general `type_id` would be handy for range of use cases (pun intended).

This is my first PR attempt to this repo - is the workflow to wait for some on t-libs to make the final call on what the fix should be?

Comment (compiler-errors) [reply library/alloc/src/collections/btree/search.rs:103]: presumably the reviewer will have a stronger opinion on this :smile_cat: though it's only been an hour, so i guess just wait for them

Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: Sounds good, thanks! :slightly_smiling_face:

Comment (the8472) [reply library/alloc/src/collections/btree/search.rs:103]: Maybe a private ZST instead of `()` would work?

Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: Feels like that should indeed buy some flexibility over `()`? Could a private ZST be "switched on" without `TypeId`/`type_id` to differentiate a map from a set?

Assumed that a `bool` or 2-variant `enum` field (both non-zero sized) would be necessary to differentiate inside a common function. But adding a field to support one error print is overkill IMO. Even "only one byte" would drag every `clone`, potentially change alignment padding, etc.

Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: Gotcha, thanks! Sounds like an excellent solution. Will wait for the t-lib reviewer's feedback before implementing.

Comment (thomcc) [reply library/alloc/src/collections/btree/search.rs:103]: I think this approach sounds like a good way forward. Do you mind trying it? Aside from that, this looks fine.

**Comment by thomcc** [comment]
I'd like to see the changes mentioned in https://github.com/rust-lang/rust/pull/98039#discussion_r899662551.

@rustbot label +S-waiting-on-author -S-waiting-on-review

File: library/alloc/src/collections/btree/search.rs (hunk 1 above)
Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: Sounds good. Traveling for the next couple days, will give it a shot ASAP.

Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: @thomcc - had a chance to try out two variants of the proposed solution. Don't believe it'll work, unfortunately.

For prototyping simplicity, am implementing the specialized variant of the blanket trait for `()` instead of introducing a new ZST.

* The same blockers would still apply for a new ZST.

* Introducing a new ZST, let's call it `SetValZST`, would have the advantage of differentiating between `BTreeMap<T, ()>` (possible, though impractical, user-defined map) and `BTreeMap<T, SetValZST>` (internal set representation). But that may be a moot point.

**Variant 1 - non-generic blanket `impl` ([here](https://github.com/tnballo/rust/commit/0397d9fb253f76297f5bbf6293cb745e3d2c1d78))**

Trait:

```rust
pub trait IsSetVal {
    fn is_set_val(&self) -> bool;
}

impl<V> IsSetVal for V {
    default fn is_set_val(&self) -> bool {
        false
    }
}

impl IsSetVal for () {
    fn is_set_val(&self) -> bool {
        true
    }
}
```

Callsite (inside function `search_tree_for_bifurcation`):

```rust
// PROBLEM: `self.node` is private. Even if zero-index is valid, cannot access instance of value.
use super::set_val::IsSetVal;
let is_set = self.node.vals[0].is_set_val();
```

Error:

```
error[E0616]: field `node` of struct `NodeRef` is private
   --> library/alloc/src/collections/btree/search.rs:102:27
    |
102 |         let is_set = self.node.vals[0].is_set_val();
    |                           ^^^^ private field
```

Cannot "reach" concrete instance of `V` to call trait function, which takes `&self` argument.

**Variant 2 - generic blanket `impl` ([here](https://github.com/tnballo/rust/commit/15b436af237669fbebb071b4cd14625b2993db6a))**

Trait:

```rust
pub trait IsSetVal<V> {
    fn is_set_val() -> bool;
}

impl<V> IsSetVal<V> for V {
    default fn is_set_val() -> bool {
        false
    }
}

// PROBLEM: conflicting overwrite of blanket impl
impl<V> IsSetVal<V> for () {
    fn is_set_val() -> bool {
        true
    }
}
```

Callsite (inside function `search_tree_for_bifurcation`):

```rust
let is_set = super::set_val::IsSetVal::<V>::is_set_val();
```

Error:

```
error[E0119]: conflicting implementations of trait `collections::btree::set_val::IsSetVal<()>` for type `()`
  --> library/alloc/src/collections/btree/set_val.rs:11:1
   |
5  | impl<V> IsSetVal<V> for V {
   | ------------------------- first implementation here
...
11 | impl<V> IsSetVal<V> for () {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `()`
```

Generic blanket implementations cannot be overwritten.

**Conclusion**

Unless there's a (specialization?) fix applicable here, think we're back to square one. Either of:

1. Truncating the message so it's always right by not mentioning type.

2. Going with the current (will incorrectly print `BTreeSet` only if `BTreeMap` value is a ZST,  theoretically that's almost never).

Googling around, [this StackOverflow post](https://stackoverflow.com/questions/60138397/how-to-test-for-type-equality-in-rust) might be helpful but it requires unstable `#![feature(specialization)]`.

Comment (the8472) [reply library/alloc/src/collections/btree/search.rs:103]: > IsSetVal<V>

This is the problem I think. The trait itself doesn't need a generic parameter.

>  `super::set_val::IsSetVal::<V>::is_set_val();`

you can do `<V as IsSetVal>::is_set_val()` instead.

Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs:103]: > This is the problem I think. The trait itself doesn't need a generic parameter.

You're right, thanks! `<V as IsSetVal>::is_set_val()` works, will document and merge into this PR's branch.

Comment (thomcc) [reply library/alloc/src/collections/btree/search.rs]: I think we should also move to using a private ZST rather than (), so that it doesn't output `BTreeSet` for `BTreeMap<K, ()>`.

But perhaps that's a bigger change than we want, and should be done later.

**Comment by thomcc** [comment]
Can you rebase rather than merge? Or squash the commits? I think we try to avoid merge commits in the history. Other than that this looks great.

File: library/alloc/src/collections/btree/search.rs (hunk 1 above)
Comment (tnballo) [reply library/alloc/src/collections/btree/search.rs]: Agreed! Was just working on that - called the private ZST `SetValZST` and added a unit test for the `BTreeMap<K, ()>` edge case.

Making the change seemed straightforward so I've added it to this PR and merged in the latest `rust-lang/rust:master` :+1:

**Comment by thomcc** [comment]
Looks great, thanks for working through this!

@bors r+
